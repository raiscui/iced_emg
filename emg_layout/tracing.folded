ThreadId(2)-tests::it_works 2756700
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 1229700
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::new_root:emg_layout\src\lib.rs:424 915300
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 1148100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout:: build new child :emg_layout\src\lib.rs:509 754600
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout:: build new child :emg_layout\src\lib.rs:509; emg_layout::calc::layout_calculating:emg_layout\src\calc.rs:29 140300
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout:: build new child :emg_layout\src\lib.rs:509 794700
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 931100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout:: build new child :emg_layout\src\lib.rs:509 576900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout:: build new child :emg_layout\src\lib.rs:509; emg_layout::calc::layout_calculating:emg_layout\src\calc.rs:29 54200
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout:: build new child :emg_layout\src\lib.rs:509 387500
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 623200
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::tests::debug print ec:emg_layout\src\lib.rs:708 10297400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 499100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Vec<Box<(dyn RefreshFor<EdgeItem> + 'static)>>:emg_layout\src\impl_refresh.rs:39 606300
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Vec<Box<(dyn RefreshFor<EdgeItem> + 'static)>>:emg_layout\src\impl_refresh.rs:39; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 1337500
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Vec<Box<(dyn RefreshFor<EdgeItem> + 'static)>>:emg_layout\src\impl_refresh.rs:39; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109; emg_layout::impl_refresh::-> css_refresh_edgedata:emg_layout\src\impl_refresh.rs:62 1758500
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Vec<Box<(dyn RefreshFor<EdgeItem> + 'static)>>:emg_layout\src\impl_refresh.rs:39; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 419400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Vec<Box<(dyn RefreshFor<EdgeItem> + 'static)>>:emg_layout\src\impl_refresh.rs:39 343000
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 12811400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 334400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109; emg_layout::impl_refresh::-> css_refresh_edgedata:emg_layout\src\impl_refresh.rs:62 658600
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 345700
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 656000
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ calculated_align ] recalculation..(&p_calculated.size, &layout.align.watch()).map :emg_layout\src\calc.rs:74 277300
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 633800
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ calculated_size ] recalculation..(&p_calculated.size, &layout.size.watch()).map :emg_layout\src\calc.rs:56 292000
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 722600
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ coordinates_trans ] recalculation..(&calculated_origin, &calculated_align).map :emg_layout\src\calc.rs:85 41900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 846500
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 246100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109; emg_layout::impl_refresh::-> css_refresh_edgedata:emg_layout\src\impl_refresh.rs:62 608900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 272300
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 578400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ calculated_size ] recalculation..(&p_calculated.size, &layout.size.watch()).map :emg_layout\src\calc.rs:56 297900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 651100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ calculated_origin ] recalculation..(&calculated_size, &layout.origin.watch()).map :emg_layout\src\calc.rs:65 293100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 650100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ coordinates_trans ] recalculation..(&calculated_origin, &calculated_align).map :emg_layout\src\calc.rs:85 113600
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 3486200
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ loc_styles ] recalculation..(&calculated_size, &matrix).map :emg_layout\src\calc.rs:107 332900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 683900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::-> [ styles ] recalculation..(&other_styles_watch, &loc_styles).map :emg_layout\src\lib.rs:522 78200
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 693900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 340800
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109; emg_layout::impl_refresh::-> css_refresh_edgedata:emg_layout\src\impl_refresh.rs:62 527000
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::impl_refresh::-> RefreshFor<EdgeItem> for Css<Use>:emg_layout\src\impl_refresh.rs:109 334500
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 630100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::p_calculated recalculation:emg_layout\src\calc.rs:31 607400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 685700
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ calculated_size ] recalculation..(&p_calculated.size, &layout.size.watch()).map :emg_layout\src\calc.rs:56 330700
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 1245100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ calculated_align ] recalculation..(&p_calculated.size, &layout.align.watch()).map :emg_layout\src\calc.rs:74 328400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 675900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ calculated_origin ] recalculation..(&calculated_size, &layout.origin.watch()).map :emg_layout\src\calc.rs:65 303100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 617600
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::calc::-> [ coordinates_trans ] recalculation..(&calculated_origin, &calculated_align).map :emg_layout\src\calc.rs:85 36800
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 782900
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::tests::debug print 1:emg_layout\src\lib.rs:776 14332100
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 3990400
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677; emg_layout::tests::debug print 2:emg_layout\src\lib.rs:781 41348300
ThreadId(2)-tests::it_works; emg_layout::tests::start:emg_layout\src\lib.rs:677 3188200
