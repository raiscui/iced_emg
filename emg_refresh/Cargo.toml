[package]
name = "emg_refresh"
version = "0.1.0"
authors = ["崔路明 <cuiluming@chaoway.net>"]
edition = "2021"
resolver = "2"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[profile.dev]
split-debuginfo = "unpacked"
[lib]
crate-type = ["cdylib", "rlib"]

[dependencies]
emg_state = { path = "../emg_state" }
emg_core = { path = "../emg_core" }

tracing = "0.1"

[target.'cfg(target_arch = "wasm32")'.dependencies]
# console_log = { version = "0.2", features = ["color"] }


[dev-dependencies]
wasm-bindgen-test = "0.3"
anchors = { path = "../../anchors" }
console_error_panic_hook = "0.1"
cfg-if = "1"
# ctor = "0.1.20"
[target.'cfg(target_arch = "wasm32")'.dev-dependencies]
tracing-wasm = "0.2.0"
