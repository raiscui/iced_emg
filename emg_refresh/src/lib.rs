#![deny(clippy::all)]
#![deny(clippy::pedantic)]
#![warn(clippy::nursery)]
// ────────────────────────────────────────────────────────────────────────────────
// #![feature(specialization)]
#![feature(min_specialization)]
#![feature(auto_traits)]
#![feature(negative_impls)]
// #![feature(trivial_bounds)]
#[cfg(test)]
mod test;
// ────────────────────────────────────────────────────────────────────────────────

mod impl_refresh;
mod refresh_use;
mod refresh_use_for;
mod refreshers;
pub use impl_refresh::RefreshUseNoWarper;
pub use impl_refresh::RefreshWhoNoWarper;
pub use refresh_use::RefreshUse;
pub use refresh_use_for::RefreshForUse;
pub use refreshers::EqRefreshFor;
pub use refreshers::RefreshFor;
pub use refreshers::Refresher;
pub use refreshers::RefresherFor;

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        let x = 2 + 2;
        assert_eq!(x, 4);
    }
}
