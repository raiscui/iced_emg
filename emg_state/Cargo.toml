[package]
name = "emg_state"
version = "0.1.0"
authors = ["崔路明 <cuiluming@chaoway.net>"]
edition = "2018"
resolver = "2"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[profile.dev]
split-debuginfo = "unpacked"

[dependencies]
anchors = { path="../../anchors" }
anymap = "0.12.1"
slotmap = "1.0.3"
topo = { version="0.13.2", features=["wasm-bindgen"] }
# log = "0.4"
derive_more = "0.99"
tracing = { version="^0.1" }
emg_core = { path="../emg_core" }
[dev-dependencies]
wasm-bindgen-test = "0.3"
