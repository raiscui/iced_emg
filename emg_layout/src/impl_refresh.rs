/*
 * @Author: Rais
 * @Date: 2021-03-29 19:22:19
 * @LastEditTime: 2022-01-07 13:08:30
 * @LastEditors: Rais
 * @Description:
 */

use emg_refresh::{RefreshFor, RefreshForUse, RefreshUseNoWarper, RefreshWhoNoWarper};
use std::{any::Any, panic::Location, rc::Rc};

use emg_state::{CloneStateVar, StateAnchor, StateVar};
pub use seed_styles as styles;
use styles::{CssHeight, CssValueTrait, CssWidth, UpdateStyle};
use tracing::{debug, trace, trace_span, warn};

use crate::{
    add_values::{AlignX, AlignY, OriginX, OriginY},
    animation::AnimationE,
    Css, EPath, EmgEdgeItem,
};

// ────────────────────────────────────────────────────────────────────────────────

//TODO lifetime
impl<Ix> RefreshWhoNoWarper for EmgEdgeItem<Ix> where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default
{
}

impl<T> RefreshUseNoWarper for Css<T> where T: CssValueTrait + Clone + 'static {}

impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for Box<dyn RefreshFor<EmgEdgeItem<Ix>>>
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
    EmgEdgeItem<Ix>: RefreshWhoNoWarper,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!(
            "!!!!!!!!!!!!!!-> RefreshFor<EdgeItem> for Box<(dyn RefreshFor<EdgeItem> + 'static)>"
        )
        .entered();
        // let ii = i.as_ref();
        who.refresh_for_use(self.as_ref());
    }
}
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for Rc<dyn RefreshFor<EmgEdgeItem<Ix>>>
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
    EmgEdgeItem<Ix>: RefreshWhoNoWarper,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!(
            "!!!!!!!!!!!!!!-> RefreshFor<EdgeItem> for Box<(dyn RefreshFor<EdgeItem> + 'static)>"
        )
        .entered();
        // let ii = i.as_ref();
        who.refresh_for_use(self.as_ref());
    }
}

impl<Ix, Use> RefreshFor<EmgEdgeItem<Ix>> for StateVar<Use>
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
    EmgEdgeItem<Ix>: RefreshWhoNoWarper,
    Use: RefreshUseNoWarper + RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
{
    // #[allow(clippy::redundant_closure_for_method_calls)]
    default fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let rc_v = self.get_var_with(emg_state::Var::get);
        warn!("Edge  Refresh use StateVar current value");
        who.refresh_for_use(&*rc_v);
    }
}
// ────────────────────────────────────────────────────────────────────────────────
// ────────────────────────────────────────────────────────────────────────────────
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for StateVar<CssWidth>
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
    EmgEdgeItem<Ix>: RefreshWhoNoWarper,
{
    #[allow(clippy::redundant_closure_for_method_calls)]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        warn!("Edge  Refresh use StateVar<CssWidth>");

        who.layout.w.set(self.watch().into());

        // who.refresh_use(&*rc_var);
    }
}
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for StateAnchor<CssWidth>
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
    EmgEdgeItem<Ix>: RefreshWhoNoWarper,
    // Use: RefreshUseNoWarper + RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
{
    #[allow(clippy::redundant_closure_for_method_calls)]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        warn!("Edge  Refresh use StateAnchor<CssWidth>");

        who.layout.w.set(self.clone().into());

        // who.refresh_use(&*rc_var);
    }
}
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for StateVar<CssHeight>
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
    EmgEdgeItem<Ix>: RefreshWhoNoWarper,
    // Use: RefreshUseNoWarper + RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
{
    #[allow(clippy::redundant_closure_for_method_calls)]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        warn!("Edge  Refresh use StateVar<CssHeight>");

        who.layout.h.set(self.watch().into());

        // who.refresh_use(&*rc_var);
    }
}
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for StateAnchor<CssHeight>
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
    EmgEdgeItem<Ix>: RefreshWhoNoWarper,
    // Use: RefreshUseNoWarper + RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
{
    #[allow(clippy::redundant_closure_for_method_calls)]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        warn!("Edge  Refresh use StateAnchor<CssHeight>");

        who.layout.h.set(self.clone().into());

        // who.refresh_use(&*rc_var);
    }
}

// ────────────────────────────────────────────────────────────────────────────────
// ────────────────────────────────────────────────────────────────────────────────

// impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for Vec<Box<(dyn RefreshFor<EmgEdgeItem<Ix>> + 'static)>>
// where
//     Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
// {
//     #[track_caller]
//     fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
//         for i in self {
//             let _g = trace_span!(
//                 "-> RefreshFor<EdgeItem> for Vec<Box<(dyn RefreshFor<EdgeItem> + 'static)>>"
//             )
//             .entered();
//             // let ii = i.as_ref();
//             who.refresh_use(i.as_ref());
//         }
//     }
// }
// impl RefreshFor<EdgeData> for Vec<Box<(dyn RefreshFor<EdgeData> + 'static)>> {
//     #[track_caller]
//     fn refresh_for(&self, _who: &mut EdgeData) {
//         panic!("!!!!!!");
//         // for i in self {
//         //     // let ii = i.as_ref();
//         //     who.refresh_use(i.as_ref());
//         // }
//     }
// }
//TODO 做 不是refresh 版本的
#[track_caller]
fn css_refresh_edgedata<Use, Ix>(css: &Css<Use>, ed: &mut EmgEdgeItem<Ix>)
where
    Use: CssValueTrait + std::clone::Clone,
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    let _g = trace_span!("-> css_refresh_edgedata").entered();

    let any = &css.0 as &dyn Any;
    if let Some(css_width) = any.downcast_ref::<CssWidth>() {
        debug!("dyn match CssWidth {}", &css_width);
        ed.refresh_for_use(css_width);
        return;
    }

    if let Some(css_height) = any.downcast_ref::<CssHeight>() {
        debug!("dyn match CssHeight {}", &css_height);
        ed.refresh_for_use(css_height);
        return;
    }

    {
        // @ 不唯一, 多次会重复 ─────────────────────────────────────────────────────────────────

        ed.other_styles.set_with(|s| {
            let mut tmp_s = s.clone();
            let t = css.0.clone();

            tmp_s
                .updated_at
                .push_back(format!("{}", Location::caller()));

            t.update_style(&mut tmp_s);
            tmp_s
        });
    }
}
// ────────────────────────────────────────────────────────────────────────────────

impl<Use, Ix> RefreshFor<EmgEdgeItem<Ix>> for Css<Use>
where
    Use: CssValueTrait + std::clone::Clone,
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!("-> RefreshFor<EdgeItem> for Css<Use>").entered();

        css_refresh_edgedata(self, who);
    }
}

// ────────────────────────────────────────────────────────────────────────────────

impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for CssWidth
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!("-> RefreshFor<EmgEdgeItem> for CssWidth").entered();

        who.layout.w.set(self.clone().into());
    }
}

impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for CssHeight
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!("-> RefreshFor<EmgEdgeItem> for CssHeight").entered();

        who.layout.h.set(self.clone().into());
    }
}
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for OriginX
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!("-> RefreshFor<EmgEdgeItem> for OriginX").entered();

        who.layout.origin_x.set(self.clone().into());
    }
}
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for OriginY
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!("-> RefreshFor<EmgEdgeItem> for OriginY").entered();

        who.layout.origin_y.set(self.clone().into());
    }
}

impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for AlignX
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!("-> RefreshFor<EmgEdgeItem> for AlignX").entered();

        who.layout.align_x.set(self.clone().into());
    }
}
impl<Ix> RefreshFor<EmgEdgeItem<Ix>> for AlignY
where
    Ix: Clone + std::hash::Hash + Eq + Ord + 'static + Default,
{
    #[track_caller]
    fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
        let _g = trace_span!("-> RefreshFor<EmgEdgeItem> for AlignY").entered();

        who.layout.align_y.set(self.clone().into());
    }
}

// ────────────────────────────────────────────────────────────────────────────────

// #[derive(Clone)]
// pub struct EffectingPath<Ix, T>(T, PhantomData<Ix>)
// where
//     T: RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
//     Ix: Clone
//         + std::hash::Hash
//         + Eq
//         + Ord
//         + 'static
//         + Default
//         + std::fmt::Debug
//         + std::fmt::Display;

// impl<Ix, T> std::ops::Deref for EffectingPath<Ix, T>
// where
//     T: RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
//     Ix: Clone
//         + std::hash::Hash
//         + Eq
//         + Ord
//         + 'static
//         + Default
//         + std::fmt::Debug
//         + std::fmt::Display,
// {
//     type Target = T;

//     fn deref(&self) -> &Self::Target {
//         &self.0
//     }
// }

// impl<T, Ix> From<T> for EffectingPath<Ix, T>
// where
//     T: RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
//     Ix: Clone
//         + std::hash::Hash
//         + Eq
//         + Ord
//         + 'static
//         + Default
//         + std::fmt::Debug
//         + std::fmt::Display,
// {
//     fn from(v: T) -> Self {
//         Self(v, PhantomData)
//     }
// }
// impl<Ix, T> RefreshFor<EmgEdgeItem<Ix>> for EffectingPath<Ix, T>
// where
//     T: RefreshFor<EmgEdgeItem<Ix>> + Clone + 'static,
//     Ix: Clone
//         + std::hash::Hash
//         + Eq
//         + Ord
//         + 'static
//         + Default
//         + std::fmt::Debug
//         + std::fmt::Display,
// {
//     fn refresh_for(&self, who: &mut EmgEdgeItem<Ix>) {
//         // self.effect_with_path(p,who);
//         // who.refresh_use(self);
//     }
// }
/// using at tree building
impl<Ix, Message> RefreshFor<EmgEdgeItem<Ix>> for AnimationE<Message>
where
    Message: Clone + std::fmt::Debug + 'static + PartialEq,
    Ix: Clone
        + std::hash::Hash
        + Eq
        + Ord
        + 'static
        + Default
        + std::fmt::Debug
        + std::fmt::Display,
{
    fn refresh_for(&self, edge: &mut EmgEdgeItem<Ix>) {
        //NOTE 当 tree 宏 中 在 edge中使用 am类型
        trace!(
            "AnimationE  RefreshFor EmgEdgeItem snapshot: \n{:#?}",
            illicit::Snapshot::get()
        );
        if let Ok(path) = illicit::get::<EPath<Ix>>() {
            let p = (*path).clone();
            self.effecting_edge_path(&*edge, p);
        }
    }
}

#[cfg(test)]
mod refresh_test {
    use std::time::Duration;

    use emg::edge_index_no_source;
    use emg::node_index;
    use emg_animation::to;
    use emg_core::vector;
    use emg_core::{into_vector, IdStr};
    use emg_refresh::RefreshForUse;
    use emg_state::CloneStateVar;
    use emg_state::{use_state, Dict, StateVar};
    use seed_styles as styles;
    use seed_styles::CssWidth;

    use styles::px;

    #[allow(unused)]
    use styles::{pc, width};

    use crate::animation::global_clock;
    use crate::EPath;
    use crate::GraphEdgesDict;
    use crate::{anima, AnimationE, EmgEdgeItem};

    #[allow(dead_code)]
    #[derive(Debug, Clone, PartialEq)]
    enum Message {
        A,
    }

    #[test]
    fn edge() {
        let e_dict_sv: StateVar<GraphEdgesDict<IdStr>> = use_state(Dict::new());
        let root_e_source = use_state(None);
        let root_e_target = use_state(Some(node_index("root")));
        let mut root_e = EmgEdgeItem::default_with_wh_in_topo(
            root_e_source.watch(),
            root_e_target.watch(),
            e_dict_sv.watch(),
            1920,
            1080,
        );

        let css_w: StateVar<CssWidth> = use_state(width(px(99)));
        let a: AnimationE<Message> = anima![css_w];
        illicit::Layer::new()
            .offer(EPath::<String>(vector![edge_index_no_source("root")]))
            .enter(|| {
                root_e.refresh_for_use(&a);
                // root_e.refresh_use(&a);
            });

        let now = global_clock();

        a.interrupt(vector![
            to(into_vector![width(px(0))]),
            to(into_vector![width(px(1))])
        ]);

        now.set(Duration::from_millis(16));
        insta::assert_debug_snapshot!("anima_refresh_edge_16", &a);
        insta::assert_debug_snapshot!("anima_refresh_edge_16_edge", &root_e);
        now.set(Duration::from_millis(33));
        insta::assert_debug_snapshot!("anima_refresh_edge_33", &a);
        insta::assert_debug_snapshot!("anima_refresh_edge_33_edge", &root_e);

        // a.effecting_edge_path(&root_e, EPath(vector![edge_index_no_source("root")]));

        // let mut pe = PathEItem(EPath(vector![edge_index_no_source("root")]), root_e);
        // bb.refresh_for(&mut pe.1);
        // bb.refresh_for(&mut pe);
        // let fff = bbb.as_ref();
        // pe.refresh_use(fff);
    }
}
