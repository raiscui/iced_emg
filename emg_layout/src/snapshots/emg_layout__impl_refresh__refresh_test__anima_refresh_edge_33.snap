---
source: emg_layout/src/impl_refresh.rs
assertion_line: 447
expression: "&a"

---
AnimationEdge {
    inside: AnimationInside {
        steps: StateVar(
            [
                _Step,
                To(
                    [
                        Prop(
                            TypeName(
                                "CssWidth",
                            ),
                            Motion {
                                position: NotNan(
                                    1.0,
                                ),
                                velocity: NotNan(
                                    0.0,
                                ),
                                target: NotNan(
                                    1.0,
                                ),
                                interpolation: Spring {
                                    stiffness: NotNan(
                                        170.0,
                                    ),
                                    damping: NotNan(
                                        26.0,
                                    ),
                                },
                                unit: Px,
                                interpolation_override: None,
                            },
                        ),
                    ],
                ),
            ],
        ),
        interruption: StateVar(
            [],
        ),
        props: [
            StateVarProperty(
                StateVar(
                    Prop(
                        TypeName(
                            "CssWidth",
                        ),
                        Motion {
                            position: NotNan(
                                87.0304518011875,
                            ),
                            velocity: NotNan(
                                -429.1130700783828,
                            ),
                            target: NotNan(
                                0.0,
                            ),
                            interpolation: Spring {
                                stiffness: NotNan(
                                    170.0,
                                ),
                                damping: NotNan(
                                    26.0,
                                ),
                            },
                            unit: Px,
                            interpolation_override: None,
                        },
                    ),
                ),
            ),
        ],
    },
    timing: StateAnchor(
        Timing {
            current: 33ms,
            dt: 17ms,
        },
    ),
    running: StateAnchor(
        true,
    ),
}
